README
Snort 2.9 introduces the DAQ, or Data Acquisition library, for packet I/O.  The
DAQ replaces direct calls to libpcap functions with an abstraction layer that
facilitates operation on a variety of hardware and software interfaces without
requiring changes to Snort.  It is possible to select the DAQ type and mode
when invoking Snort to perform pcap readback or inline operation, etc.  The
DAQ library may be useful for other packet processing applications and the
modular nature allows you to build new modules for other platforms.

This README summarizes the important things you need to know to use the DAQ.


Building the DAQ Library and DAQ Modules
========================================

The DAQ is bundled with Snort but must be built first using these steps:

    ./configure
    make
    sudo make install

This will build and install both static and dynamic DAQ modules.

Note that pcap >= 1.0.0 is required.  pcap 1.1.1 is available at the time
of this writing and is recommended.

Also, libdnet is required for IPQ and NFQ DAQs.  If you get a relocation error
trying to build those DAQs, you may need to reinstall libdnet and configure it
with something like this:

    ./configure "CFLAGS=-fPIC -g -O2"

You may also experience problems trying to find the dynamic dnet library
because it isn't always named properly.  Try creating a link to the shared
library (identified by its .x or .x.y etc. extension) with the same name but
with ".so" inserted as follows:

    $ ln -s libdnet.1.1 libdnet.so.1.1
    $ ldconfig -Rv /usr/local/lib 2>&1 | grep dnet
      Adding /usr/local/lib/libdnet.so.1.1

Alternatively, you should be able to fix both issues as follows:

    libtoolize --copy --force
    aclocal -I config
    autoheader
    autoconf
    automake --foreign

When the DAQ library is built, both static and dynamic flavors will be
generated.  The various DAQ modules will be built if the requisite headers and
libraries are available.  You can disable individual modules, etc. with options
to configure.  For the complete list of configure options, run:

    ./configure --help


PCAP Module
===========

pcap is the default DAQ.  If snort is run w/o any DAQ arguments, it will
operate as it always did using this module.  These are equivalent:

    ./snort -i <device>
    ./snort -r <file>

    ./snort --daq pcap --daq-mode passive -i <device>
    ./snort --daq pcap --daq-mode read-file -r <file>

You can specify the buffer size pcap uses with:

    ./snort --daq pcap --daq-var buffer_size=<#bytes>

* The pcap DAQ does not count filtered packets. *


AFPACKET Module
===============

afpacket functions similar to the pcap DAQ but with better performance:

    ./snort --daq afpacket -i <device>
            [--daq-var buffer_size_mb=<#MB>]
            [--daq-var debug]

If you want to run afpacket in inline mode, you must set device to one or more
interface pairs, where each member of a pair is separated by a single colon and
each pair is separated by a double colon like this:

    eth0:eth1

or this:

    eth0:eth1::eth2:eth3

By default, the afpacket DAQ allocates 128MB for packet memory.  You can change
this with:

    --daq-var buffer_size_mb=<#MB>

Note that the total allocated is actually higher, here's why.  Assuming the
default packet memory with a snaplen of 1518, the numbers break down like this:

* The frame size is 1518 (snaplen) + the size of the AFPacket header (66
  bytes) = 1584 bytes.

* The number of frames is 128 MB / 1518 = 84733.

* The smallest block size that can fit at least one frame is  4 KB = 4096 bytes
  @ 2 frames per block.

* As a result, we need 84733 / 2 = 42366 blocks.

* Actual memory allocated is 42366 * 4 KB = 165.5 MB.


NFQ Module
==========

NFQ is the new and improved way to process iptables packets:

    ./snort --daq nfq \
        [--daq-var device=<dev>] \
        [--daq-var proto=<proto>] \
        [--daq-var queue=<qid>]

    <dev> ::= ip | eth0, etc; default is IP injection
    <proto> ::= ip4 | ip6 |; default is ip4
    <qid> ::= 0..65535; default is 0

This module can not run unprivileged so ./snort -u -g will produce a warning
and won't change user or group.

Notes on iptables are given below.


IPQ Module
==========

IPQ is the old way to process iptables packets.  It replaces the inline version
available in pre-2.9 versions built with this:

    ./configure --enable-inline

Note that layer 2 resets are not supported with the IPQ DAQ:

    config layer2resets[: <mac>]

Start the IPQ DAQ as follows:

    ./snort --daq ipq \
        [--daq-var device=<dev>] \
        [--daq-var proto=<proto>] \

    <dev> ::= ip | eth0, etc; default is IP injection
    <proto> ::= ip4 | ip6; default is ip4

This module can not run unprivileged so ./snort -u -g will produce a warning
and won't change user or group.

Notes on iptables are given below.


IPFW Module
===========

IPFW is available for BSD systems.  It replaces the inline version available in
pre-2.9 versions built with this:

    ./configure --enable-ipfw

This command line argument is no longer supported:

    ./snort -J <port#>

Instead, start Snort like this:

    ./snort --daq ipfw [--daq-var port=<port>]

    <port> ::= 1..65535; default is 8000

* IPFW only supports ip4 traffic.

Notes on FreeBSD and OpenBSD are given below.


Dump Module
===========

The dump DAQ allows you to test the various inline mode features available in
2.9 Snort like injection and normalization.

    ./snort -i <device> --daq dump
    ./snort -r <pcap> --daq dump

By default a file named inline-out.pcap will be created containing all packets
that passed through or were generated by snort.  You can optionally specify a
different name.

    ./snort --daq dump --daq-var file=<name>

dump uses the pcap daq for packet acquisition.  It therefore does not count
filtered packets (a pcap limitation).

Note that the dump DAQ inline mode is not an actual inline mode.  Furthermore,
you will probably want to have the pcap DAQ acquire in another mode like this:

    ./snort -r <pcap> -Q --daq dump --daq-var load-mode=read-file
    ./snort -i <device> -Q --daq dump --daq-var load-mode=passive


Notes on iptables
=================

These notes are just a quick reminder that you need to set up iptables to use
the IPQ or NFQ DAQs.  Doing so may cause problems with your network so tread
carefully.  The examples below are intentionally incomplete so please read the
related documentation first.

Here is a blog post by Marty for historical reference:

    http://archives.neohapsis.com/archives/snort/2000-11/0394.html

You can check this out for queue sizing tips:

    http://www.inliniac.net/blog/2008/01/23/improving-snort_inlines-nfq-performance.html

You might find useful IPQ info here:

    http://snort-inline.sourceforge.net/

Use this to examine your iptables:

    sudo /sbin/iptables -L

Use something like this to set up NFQ:

    sudo /sbin/iptables
        -I <table> [<protocol stuff>] [<state stuff>]
        -j NFQUEUE --queue-num 1

Use something like this to set up IPQ:

    sudo iptables -I FORWARD -j QUEUE

Use something like this to "disconnect" snort:

    sudo /sbin/iptables -D <table> <rule pos>

Be sure to start Snort prior to routing packets through NFQ with iptables.
Such packets will be dropped until Snort is started.

The queue-num is the number you must give Snort.

If you are running on a system with both NFQ and IPQ support, you may
experience some start-up failures of the sort:

The solution seems to be to remove both modules from the kernel like this:

    modprobe -r nfnetlink_queue
    modprobe -r ip_queue

and then install the module you want:

    modprobe ip_queue

or:

    modprobe nfnetlink_queue

These DAQs should be run with a snaplen of 65535 since the kernel defrags the
packets before queuing.  Also, no need to configure frag3.


Notes on FreeBSD::IPFW
======================

Check the online manual at:

    http://www.freebsd.org/doc/handbook/firewalls-ipfw.html.

Here is a brief example to divert icmp packets to Snort at port 8000:

To enable support for divert sockets, place the following lines in the
kernel configuration file:

    options IPFIREWALL
    options IPDIVERT

(The file in this case was: /usr/src/sys/i386/conf/GENERIC; which is platform
dependent.)

You may need to also set these to use the loadable kernel modules:

/etc/rc.conf:
firewall_enable="YES"

/boot/loader.conf:
ipfw_load="YES"
ipdivert_load="YES"

$ dmesg | grep ipfw
ipfw2 (+ipv6) initialized, divert loadable, nat loadable, rule-based

ChangeLog
Changes in 2.0.1 Released on 2013-07-22:
----------------------------------------
2013-06-28 11:00  rcombs

    * daq.h, daq_api.h, daq_base.c, daq_common.h, daq_mod_ops.c, 
      daq_afpacket.c, daq_dump.c, daq_ipfw.c, daq_ipq.c, daq_nfq.c, 
      daq_pcap.c, daq_static_modules.c, daq_static_modules.h, 
      sf_bpf_filter.c, sf_bpf_printer.c, sf_gencode.c, sf_nametoaddr.c, 
      sf_optimize.c, sfbpf-int.c, sfbpf-int.h, sfbpf.h, 
      sfbpf_dlt.h:
      Update copyright year.

    * daq_dump.c, daq_ipfw.c, daq_ipq.c, daq_nfq.c: 
      Ensure verdict is in range before bumping peg counts.
      Thanks to John Menerick <omgclouds@gmail.com> for reporting the
      issue.

Changes in 2.0.0 Released on 2012-10-09:
----------------------------------------
2012-10-01 16:15  ssturges

    * daq.spec: 
      Update DAQ spec file to reflect 2.0.0.

2012-09-07 15:30  maltizer

    * sfbpf/sll.h:
      Fix sfbpf compilation on Solaris.

2012-09-06 17:56  hsugar

    * api/daq.h: 
      Add a DAQ version for conditional compiles.

2012-09-06 11:17  maltizer

    * COPYING, configure.ac, api/Makefile.am, api/daq.h, api/daq_api.h,
      api/daq_base.c, api/daq_common.h, api/daq_mod_ops.c,
      os-daq-modules/daq_afpacket.c, os-daq-modules/daq_dump.c,
      os-daq-modules/daq_ipfw.c, os-daq-modules/daq_ipq.c,
      os-daq-modules/daq_nfq.c, os-daq-modules/daq_pcap.c,
      os-daq-modules/daq_static_modules.c,
      os-daq-modules/daq_static_modules.h, sfbpf/sf_bpf_printer.c:
      Add address_space_id fields to DAQ packet headers and flow
      messages.  Rev LibDAQ to major 2.  Update the address of the FSF.

Changes in 1.1.1 Released on 2012-07-18:
----------------------------------------
2012-01-11 17:59  maltizer

    * configure.ac, daq.spec, os-daq-modules/Makefile.am:
      Fix build when configuring with --disable-shared (don't build
      dynamic modules).  Add note to AFPacket section of README warning
      against RHEL5's kernel.  Update LibDAQ version to 1.1.1.

2012-01-04 00:54  maltizer

    * api/daq_common.h:
      Fix flag definition.

2011-12-23 14:25  maltizer

    * configure.ac, daq.spec, api/daq_common.h:
      Add the DAQ_PKT_FLAG_NOT_FORWARDING flag

2011-11-04 13:28  maltizer

    * configure.ac, daq.spec, api/Makefile.am:
      Update the SO major revision of libdaq to 1.


Changes in 0.6.2 Released on 2011-10-06:
----------------------------------------
2011-09-21 16:22  rcombs

    * configure.ac, os-daq-modules/Makefile.am:
      fix dnet vs dumbnet in a more righteous way

2011-07-18 10:05  rdempster

    * sfbpf/Makefile.am:
      Remove GNU Make specific declarations and replace with srcdir
      literals.

Changes in 0.6 Released on 2011-08-23:
--------------------------------------

2011-06-08 17:26  rcombs

    * daq.dsp, daq.spec, api/daq.h, api/daq_common.h:
      convert spec file to unix flavor; fix windows build

2011-05-27 12:07  rdempster

    * configure.ac, os-daq-modules/daq-modules-config.in:
      Fix up some typos and shell errors.

2011-05-13 15:50  rdempster

    * configure.ac, daq.spec, api/daq.h, api/daq_common.h,
      m4/ax_cflags_gcc_option.m4, os-daq-modules/Makefile.am,
      os-daq-modules/daq-modules-config.in,
      os-daq-modules/daq_afpacket.c, os-daq-modules/daq_dump.c,
      os-daq-modules/daq_ipfw.c, os-daq-modules/daq_ipq.c,
      os-daq-modules/daq_nfq.c, os-daq-modules/daq_pcap.c,
      sfbpf/Makefile.am, sfbpf/sf-redefines.h, sfbpf/sf_bpf_filter.c,
      sfbpf/sf_bpf_printer.c, sfbpf/sf_gencode.c, sfbpf/sfbpf-int.h: 
      Update daq to gracefully handle dnet as dumbnet on ubuntu.
      Make libpcap export pcap_version, so that it works with the daq
      configure script.

2011-05-13 15:46  rcombs

    * os-daq-modules/daq_nfq.c:
      convert sec to ms; disable ENOBUFS

2011-02-15 17:59  maltizer

    * os-daq-modules/daq_pcap.c:
      Fix typo in pcap DAQ module's buffer_size variable parsing.

2010-12-08 11:03  rcombs

    * os-daq-modules/daq_ipfw.c:
      use ipv4 instead of raw

2010-12-08 10:53  rcombs

    * README, os-daq-modules/daq_ipq.c, os-daq-modules/daq_nfq.c:
      discriminate between ip4 and ip6 raw packets

2010-12-08 10:30  rcombs

    * os-daq-modules/Makefile.am: Bugs Fixed: 81786 Change description:

os-daq-modules/daq_nfq.c
/*
 ** Portions Copyright (C) 1998-2010 Sourcefire, Inc.
 **
 ** This program is free software; you can redistribute it and/or modify
 ** it under the terms of the GNU General Public License as published by
 ** the Free Software Foundation; either version 2 of the License, or
 ** (at your option) any later version.
 **
 ** This program is distributed in the hope that it will be useful,
 ** but WITHOUT ANY WARRANTY; without even the implied warranty of
 ** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 ** GNU General Public License for more details.
 **
 ** You should have received a copy of the GNU General Public License
 ** along with this program; if not, write to the Free Software
 ** Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <sys/types.h>
#include <sys/time.h>
#include <sys/unistd.h>

#include <netinet/ip.h>

#ifdef HAVE_DUMBNET_H
#include <dumbnet.h>
#else
#include <dnet.h>
#endif
#include <linux/netfilter.h>
#include <libnetfilter_queue/libnetfilter_queue.h>

#include "daq_api.h"
#include "sfbpf.h"

#define DAQ_MOD_VERSION  6

#define DAQ_NAME "nfq"
#define DAQ_TYPE (DAQ_TYPE_INTF_CAPABLE | DAQ_TYPE_INLINE_CAPABLE | \
                  DAQ_TYPE_MULTI_INSTANCE | DAQ_TYPE_NO_UNPRIV)

// FIXTHIS meta data is 80 bytes on my test platform but who knows?
// documentation is poor; need correct way to size meta data
// erring on the high side here to avoid truncation errors
#define META_DATA_SIZE 512

#define MSG_BUF_SIZE META_DATA_SIZE + IP_MAXPACKET

typedef struct
{
    int protos, sock, qid;
    int qlen;

    struct nfq_handle* nf_handle;
    struct nfq_q_handle* nf_queue;

    const char* device;
    char* filter;
    struct sfbpf_program fcode;

    ip_t* net;
    eth_t* link;

    uint8_t* buf;
    void* user_data;
    DAQ_Analysis_Func_t user_func;

    volatile int count;
    int passive;
    uint32_t snaplen;
    unsigned timeout;

    char error[DAQ_ERRBUF_SIZE];
    DAQ_State state;
    DAQ_Stats_t stats;
} NfqImpl;

static void nfq_daq_shutdown(void* handle);
static int daq_nfq_callback(struct nfq_q_handle* qh, struct nfgenmsg* nfmsg,
    struct nfq_data* nfad, void* data);

//-------------------------------------------------------------------------
// utilities

#define DEFAULT_Q 0

#define IP4(i) (i->protos & 0x1)
#define IP6(i) (i->protos & 0x2)

// FIXTHIS ip4 is always enabled??

static int nfq_daq_get_protos (const char* s)
{
    if ( !s || !strncasecmp(s, "ip4", 3) )
        return 0x1;

    if ( !strncasecmp(s, "ip6", 3) )
        return 0x2;
#if 0
    // doesn't look like both can be handled simultaneously
    if ( !strncasecmp(s, "ip*", 3) )
        return 0x3;
#endif
    return 0;
}

static int nfq_daq_get_setup (
    NfqImpl* impl, const DAQ_Config_t* cfg, char* errBuf, size_t errMax)
{
    DAQ_Dict* entry;

    impl->protos = 0x1;
    impl->qid = DEFAULT_Q;
    impl->qlen = 0;

    for ( entry = cfg->values; entry; entry = entry->next)
    {
        if ( !entry->value || !*entry->value )
        {
            snprintf(errBuf, errMax,
                "%s: variable needs value (%s)\n", __FUNCTION__, entry->key);
                return DAQ_ERROR;
        }
        else if ( !strcmp(entry->key, "device") )
        {
            impl->device = strdup(entry->value);
            if ( !impl->device )
            {
                snprintf(errBuf, errMax,
                    "%s: can't allocate memory for device (%s)\n",
                    __FUNCTION__, entry->value);
                return DAQ_ERROR;
            }
        }
        else if ( !strcmp(entry->key, "proto") )
        {
            impl->protos = nfq_daq_get_protos(entry->value);
            if ( !impl->protos )
            {
                snprintf(errBuf, errMax, "%s: bad proto (%s)\n",
                    __FUNCTION__, entry->value);
                return DAQ_ERROR;
            }
        }
        else if ( !strcmp(entry->key, "queue") )
        {
            char* end = entry->value;

            impl->qid = (int)strtol(entry->value, &end, 0);

            if ( *end || impl->qid < 0 || impl->qid > 65535 )
            {
                snprintf(errBuf, errMax, "%s: bad queue (%s)\n",
                    __FUNCTION__, entry->value);
                return DAQ_ERROR;
            }
        }
        else if ( !strcmp(entry->key, "queue_len") )
        {
            char* end = entry->value;
            impl->qlen = (int)strtol(entry->value, &end, 0);

            if ( *end || impl->qlen < 0 || impl->qlen > 65535 )
            {
                snprintf(errBuf, errMax, "%s: bad queue length (%s)\n",
                        __FUNCTION__, entry->value);
                return DAQ_ERROR;
            }
        }
        else
        {
            snprintf(errBuf, errMax,
                "%s: unsupported variable (%s=%s)\n",
                    __FUNCTION__, entry->key, entry->value);
                return DAQ_ERROR;
        }
    }

    impl->snaplen = cfg->snaplen ? cfg->snaplen : IP_MAXPACKET;
    impl->timeout = cfg->timeout / 1000;    // convert ms to secs
    impl->passive = ( cfg->mode == DAQ_MODE_PASSIVE );

    return DAQ_SUCCESS;
}

//-------------------------------------------------------------------------

static int nfq_daq_initialize (
    const DAQ_Config_t* cfg, void** handle, char* errBuf, size_t errMax)
{
    if(cfg->name && *(cfg->name))
    {
        snprintf(errBuf, errMax, "The nfq DAQ module does not support interface or readback mode!");
        return DAQ_ERROR_INVAL;
    }
    // setup internal stuff
    NfqImpl *impl = calloc(1, sizeof(*impl));

    if ( !impl )
    {
        snprintf(errBuf, errMax, "%s: failed to allocate nfq context\n",
            __FUNCTION__);
        return DAQ_ERROR_NOMEM;
    }

    if ( nfq_daq_get_setup(impl, cfg, errBuf, errMax) != DAQ_SUCCESS )
    {
        nfq_daq_shutdown(impl);
        return DAQ_ERROR;
    }

    if ( (impl->buf = malloc(MSG_BUF_SIZE)) == NULL )
    {
        snprintf(errBuf, errMax, "%s: failed to allocate nfq buffer\n",
            __FUNCTION__);
        nfq_daq_shutdown(impl);
        return DAQ_ERROR_NOMEM;
    }

    // setup input stuff
    // 1. get a new q handle
    if ( !(impl->nf_handle = nfq_open()) )
    {
        snprintf(errBuf, errMax, "%s: failed to get handle for nfq\n",
            __FUNCTION__);
        nfq_daq_shutdown(impl);
        return DAQ_ERROR;
    }

    // 2. now use the new q handle to rip the rug out from other
    //    nfq users / handles?  actually that doesn't seem to
    //    happen which is good, but then why is this *supposed*
    //    to be necessary?  especially since we haven't bound to
    //    a qid yet, and that is exclusive anyway.
    if (
        (IP4(impl) && nfq_unbind_pf(impl->nf_handle, PF_INET) < 0) ||
        (IP6(impl) && nfq_unbind_pf(impl->nf_handle, PF_INET6) < 0) )
    {
        snprintf(errBuf, errMax, "%s: failed to unbind protocols for nfq\n",
            __FUNCTION__);
        //nfq_daq_shutdown(impl);
        //return DAQ_ERROR;
    }

    // 3. select protocols for the q handle
    //    this is necessary but insufficient because we still
    //    must configure iptables externally, eg:
    //
    //    iptables -A OUTPUT -p icmp -j NFQUEUE [--queue-num <#>]
    //    (# defaults to 0).
    //
    // :( iptables rules should be managed automatically to avoid
    //    queueing packets to nowhere or waiting for packets that
    //    will never come.  (ie this bind should take the -p, -s,
    //    etc args you can pass to iptables and create the dang
    //    rule!)
    if (
        (IP4(impl) && nfq_bind_pf(impl->nf_handle, PF_INET) < 0) ||
        (IP6(impl) && nfq_bind_pf(impl->nf_handle, PF_INET6) < 0) )
    {
        snprintf(errBuf, errMax, "%s: failed to bind protocols for nfq\n",
            __FUNCTION__);
        nfq_daq_shutdown(impl);
        return DAQ_ERROR;
    }

    // 4. bind to/allocate the specified nfqueue instance
    //    (this is the puppy specified via iptables as in
    //    above example.)
    //
    // ** there can be at most 1 nf_queue per qid
    if ( !(impl->nf_queue = nfq_create_queue(
        impl->nf_handle, impl->qid, daq_nfq_callback, impl)) )
    {
        snprintf(errBuf, errMax, "%s: nf queue creation failed\n",
            __FUNCTION__);
        nfq_daq_shutdown(impl);
        return DAQ_ERROR;
    }

    // 5. configure copying for maximum overhead
    if ( nfq_set_mode(impl->nf_queue, NFQNL_COPY_PACKET, IP_MAXPACKET) < 0 )
    {
        snprintf(errBuf, errMax, "%s: unable to set packet copy mode\n",
            __FUNCTION__);
        nfq_daq_shutdown(impl);
        return DAQ_ERROR;
    }

    // 6. set queue length (optional)
    if ( impl->qlen > 0 &&
            nfq_set_queue_maxlen(impl->nf_queue, impl->qlen))
    {
        snprintf(errBuf, errMax, "%s: unable to set queue length\n",
                __FUNCTION__);
        nfq_daq_shutdown(impl);
        return DAQ_ERROR;
    }

    // 7. get the q socket descriptor
    //    (after getting not 1 but 2 handles!)
    impl->sock = nfq_fd(impl->nf_handle);

    // setup output stuff
    // we've got 2 handles and a socket descriptor but, incredibly,
    // no way to inject?
    if ( impl->device && strcasecmp(impl->device, "ip") )
    {
        impl->link = eth_open(impl->device);

        if ( !impl->link )
        {
            snprintf(errBuf, errMax, "%s: can't open %s!\n",
                __FUNCTION__, impl->device);
            nfq_daq_shutdown(impl);
            return DAQ_ERROR;
        }
    }
    else
    {
        impl->net = ip_open();

        if ( !impl->net )
        {
            snprintf(errBuf, errMax, "%s: can't open ip!\n", __FUNCTION__);
            nfq_daq_shutdown(impl);
            return DAQ_ERROR;
        }
    }

    // from Florign Westphal ...
    // tell kernel that we do not need to know about queue overflows.
    // Without this, read operations on the netlink socket will fail
    // with ENOBUFS on queue overrun.
    {
        int option = 1;
        setsockopt(impl->sock, SOL_NETLINK, NETLINK_NO_ENOBUFS,
                   &option, sizeof(option));
    }
    impl->state = DAQ_STATE_INITIALIZED;

    *handle = impl;
    return DAQ_SUCCESS;
}

//-------------------------------------------------------------------------

static void nfq_daq_shutdown (void* handle)
{
    NfqImpl *impl = (NfqImpl*)handle;
    impl->state = DAQ_STATE_UNINITIALIZED;

    if (impl->nf_queue)
        nfq_destroy_queue(impl->nf_queue);

    // note that we don't unbind here because
    // we will unbind other programs too

    if(impl->nf_handle)
        nfq_close(impl->nf_handle);

    if ( impl->link )
        eth_close(impl->link);

    if ( impl->net )
        ip_close(impl->net);

    if ( impl->filter )
        free(impl->filter);

    if ( impl->buf )
        free(impl->buf);

    free(impl);
}

//-------------------------------------------------------------------------

static inline int SetPktHdr (
    NfqImpl* impl,
    struct nfq_data* nfad,
    DAQ_PktHdr_t* hdr,
    uint8_t** pkt
) {
    int len = nfq_get_payload(nfad, (char**)pkt);

    if ( len <= 0 )
        return -1;

    hdr->caplen = ((uint32_t)len <= impl->snaplen) ? (uint32_t)len : impl->snaplen;
    hdr->pktlen = len;
    hdr->flags = 0;

    // if nfq fails to provide a timestamp, we fall back on tod
    if ( nfq_get_timestamp(nfad, &hdr->ts) )
        gettimeofday(&hdr->ts, NULL);

    hdr->ingress_index = nfq_get_physindev(nfad);
    hdr->egress_index = -1;
    hdr->ingress_group = -1;
    hdr->egress_group = -1;

    return 0;
}

//-------------------------------------------------------------------------

// forward all but blocks and blacklists:
static const int s_fwd[MAX_DAQ_VERDICT] = { 1, 0, 1, 1, 0, 1 };

static int daq_nfq_callback(
    struct nfq_q_handle* qh,
    struct nfgenmsg* nfmsg,
    struct nfq_data* nfad,
    void* data)
{
    NfqImpl *impl = (NfqImpl*)data;
    struct nfqnl_msg_packet_hdr* ph = nfq_get_msg_packet_hdr(nfad);

    DAQ_Verdict verdict;
    DAQ_PktHdr_t hdr;
    uint8_t* pkt;
    int nf_verdict;
    uint32_t data_len;

    if ( impl->state != DAQ_STATE_STARTED )
        return -1;

    if ( !ph || SetPktHdr(impl, nfad, &hdr, &pkt) )
    {
        DPE(impl->error, "%s: can't setup packet header",
            __FUNCTION__);
        return -1;
    }

    if (
        impl->fcode.bf_insns &&
        sfbpf_filter(impl->fcode.bf_insns, pkt, hdr.caplen, hdr.caplen) == 0
    ) {
        verdict = DAQ_VERDICT_PASS;
        impl->stats.packets_filtered++;
    }
    else
    {
        verdict = impl->user_func(impl->user_data, &hdr, pkt);
        impl->stats.verdicts[verdict]++;
        impl->stats.packets_received++;
    }
    nf_verdict = ( impl->passive || s_fwd[verdict] ) ? NF_ACCEPT : NF_DROP;
    data_len = ( verdict == DAQ_VERDICT_REPLACE ) ? hdr.caplen : 0;

    nfq_set_verdict(
        impl->nf_queue, ntohl(ph->packet_id),
        nf_verdict, data_len, pkt);

    return 0;
}

//-------------------------------------------------------------------------
// 0. we open the queue and supply our internal callback,
//    daq_nfq_callback.
// 1. the daq client calls in here to get packets.
//    we save off the user's data and callback.
// 2. then we call nfq_handle_packet for each packet received
// 3. nfq_handle_packet casts a magic spell and then calls our
//    internal callback.
// 4. that in turn applies an optional bpf and passes traffic
//    that is filtered out.
// 5. traffic that is not filtered out is passed to the previously
//    saved user callback along with the user data.
// 6. the verdict returned from the callback is used to issue
//    a pass / drop verdict to the nfq.
// 7. this unwinds and we repeat back at step 2.

static int nfq_daq_acquire (
    void* handle, int c, DAQ_Analysis_Func_t callback, DAQ_Meta_Func_t metaback, void* user)
{
    NfqImpl *impl = (NfqImpl*)handle;

    int n = 0;
    fd_set fdset;
    struct timeval tv;
    tv.tv_usec = 0;

    // If c is <= 0, don't limit the packets acquired.  However,
    // impl->count = 0 has a special meaning, so interpret accordingly.
    impl->count = (c == 0) ? -1 : c;
    impl->user_data = user;
    impl->user_func = callback;

    while ( (impl->count < 0) || (n < impl->count) )
    {
        FD_ZERO(&fdset);
        FD_SET(impl->sock, &fdset);

        // set this per call
        tv.tv_sec = impl->timeout;
        tv.tv_usec = 0;

        // at least ipq had a timeout!
        if ( select(impl->sock+1, &fdset, NULL, NULL, &tv) < 0 )
        {
            if ( errno == EINTR )
                break;
            DPE(impl->error, "%s: select = %s",
                __FUNCTION__, strerror(errno));
            return DAQ_ERROR;
        }

        if (FD_ISSET(impl->sock, &fdset))
        {
            int len = recv(impl->sock, impl->buf, MSG_BUF_SIZE, 0);

            if ( len > 0)
            {
                int stat = nfq_handle_packet(
                    impl->nf_handle, (char*)impl->buf, len);

                impl->stats.hw_packets_received++;

                if ( stat < 0 )
                {
                    DPE(impl->error, "%s: nfq_handle_packet = %s",
                        __FUNCTION__, strerror(errno));
                    return DAQ_ERROR;
                }
                n++;
            }
        }
    }
    return 0;
}

//-------------------------------------------------------------------------

static int nfq_daq_inject (
    void* handle, const DAQ_PktHdr_t* hdr, const uint8_t* buf, uint32_t len,
    int reverse)
{
    NfqImpl* impl = (NfqImpl*)handle;
    ssize_t sent = 0;

    if ( impl->link )
        sent = eth_send(impl->link, buf, len);

    else if ( impl->net )
        sent = ip_send(impl->net, buf, len);

    if ( sent != len )
    {
        DPE(impl->error, "%s: failed to send",
            __FUNCTION__);
        return DAQ_ERROR;
    }
    impl->stats.packets_injected++;
    return DAQ_SUCCESS;
}

//-------------------------------------------------------------------------

static int nfq_daq_set_filter (void* handle, const char* filter)
{
    NfqImpl* impl = (NfqImpl*)handle;
    struct sfbpf_program fcode;
    int dlt = IP4(impl) ? DLT_IPV4 : DLT_IPV6;

    if (sfbpf_compile(impl->snaplen, dlt, &fcode, filter, 1, 0) < 0)
    {
        DPE(impl->error, "%s: failed to compile bpf '%s'",
            __FUNCTION__, filter);
        return DAQ_ERROR;
    }

    if ( impl->filter )
        free((void *)impl->filter);

    if ( impl->fcode.bf_insns )
        free(impl->fcode.bf_insns);

    impl->filter = strdup(filter);
    impl->fcode = fcode;

    return DAQ_SUCCESS;
}

//-------------------------------------------------------------------------

static int nfq_daq_start (void* handle)
{
    NfqImpl* impl = (NfqImpl*)handle;
    impl->state = DAQ_STATE_STARTED;
    return DAQ_SUCCESS;
}

static int nfq_daq_breakloop (void* handle)
{
    NfqImpl* impl = (NfqImpl*)handle;
    impl->count = 0;
    return DAQ_SUCCESS;
}

static int nfq_daq_stop (void* handle)
{
    NfqImpl* impl = (NfqImpl*)handle;
    impl->state = DAQ_STATE_STOPPED;
    return DAQ_SUCCESS;
}

static DAQ_State nfq_daq_check_status (void* handle)
{
    NfqImpl* impl = (NfqImpl*)handle;
    return impl->state;
}

static int nfq_daq_get_stats (void* handle, DAQ_Stats_t* stats)
{
    NfqImpl* impl = (NfqImpl*)handle;
    *stats = impl->stats;
    return DAQ_SUCCESS;
}

static void nfq_daq_reset_stats (void* handle)
{
    NfqImpl* impl = (NfqImpl*)handle;
    memset(&impl->stats, 0, sizeof(impl->stats));
}

static int nfq_daq_get_snaplen (void* handle)
{
    NfqImpl* impl = (NfqImpl*)handle;
    return impl->snaplen;
}

static uint32_t nfq_daq_get_capabilities (void* handle)
{
    NfqImpl* impl = (NfqImpl*)handle;
    uint32_t caps = DAQ_CAPA_BLOCK | DAQ_CAPA_REPLACE | DAQ_CAPA_INJECT
        | DAQ_CAPA_BREAKLOOP | DAQ_CAPA_UNPRIV_START | DAQ_CAPA_BPF;
    if ( impl->net ) caps |= DAQ_CAPA_INJECT_RAW;
    return caps;
}

static int nfq_daq_get_datalink_type(void *handle)
{
    NfqImpl* impl = (NfqImpl*)handle;
    int dlt = IP4(impl) ? DLT_IPV4 : DLT_IPV6;
    return dlt;
}

static const char* nfq_daq_get_errbuf (void* handle)
{
    NfqImpl* impl = (NfqImpl*)handle;
    return (char*)impl->error;
}

static void nfq_daq_set_errbuf (void* handle, const char* s)
{
    NfqImpl* impl = (NfqImpl*)handle;
    DPE(impl->error, "%s", s ? s : "");
}

static int nfq_daq_get_device_index(void* handle, const char* device)
{
    return DAQ_ERROR_NOTSUP;
}

//-------------------------------------------------------------------------

#ifdef BUILDING_SO
DAQ_SO_PUBLIC DAQ_Module_t DAQ_MODULE_DATA =
#else
DAQ_Module_t nfq_daq_module_data =
#endif
{
    .api_version = DAQ_API_VERSION,
    .module_version = DAQ_MOD_VERSION,
    .name = DAQ_NAME,
    .type = DAQ_TYPE,
    .initialize = nfq_daq_initialize,
    .set_filter = nfq_daq_set_filter,
    .start = nfq_daq_start,
    .acquire = nfq_daq_acquire,
    .inject = nfq_daq_inject,
    .breakloop = nfq_daq_breakloop,
    .stop = nfq_daq_stop,
    .shutdown = nfq_daq_shutdown,
    .check_status = nfq_daq_check_status,
    .get_stats = nfq_daq_get_stats,
    .reset_stats = nfq_daq_reset_stats,
    .get_snaplen = nfq_daq_get_snaplen,
    .get_capabilities = nfq_daq_get_capabilities,
    .get_datalink_type = nfq_daq_get_datalink_type,
    .get_errbuf = nfq_daq_get_errbuf,
    .set_errbuf = nfq_daq_set_errbuf,
    .get_device_index = nfq_daq_get_device_index,
    .modify_flow = NULL,
    .hup_prep = NULL,
    .hup_apply = NULL,
    .hup_post = NULL,
};

